dali/ft/ftbase.cpp

```diff
@@ -556,18 +558,21 @@ MemoryBuffer & OutputProgress::deserializeCore(MemoryBuffer & in)
     return in;
 }
 
-MemoryBuffer & OutputProgress::deserializeExtra(MemoryBuffer & in, unsigned version)
+MemoryBuffer & OutputProgress::deserializeExtra(MemoryBuffer & in, unsigned & version)
 {
-    if (in.remaining())
+    if (version>=1 && in.remaining())
     {
-        switch (version)
-        {
-        case 1:
-            in.read(hasCompressed);
-            if (hasCompressed)
-                in.read(compressedPartSize);
-            break;
-        }
+        in.read(hasCompressed);
+        if (hasCompressed && in.remaining())
+            in.read(compressedPartSize);
+    }
+    if (version>=2)
```

Comment:
This is a bit inconsistent with serializeExtra.  SerializeExtra serialises either version 1 or version 2.  This serialises either version 1 or version 2.  I suspect clearer to match the serialize function, and require an extra call (and remove the verison modifying).

Reply:
agree, don't think it is worth the reference/changing it.
If there's anything remaining after version 1 deserialization, then it has to be >=2 and it is guaranteed to be compatible with current version. So this code can unconditionally deserialize numWrites/numReads if more remaining() (as guaranteed to be at least ver. 2).
If anything else is wanted in future, it can conditionally deserialize based on version.

Reply:
I can't assume that any remaining data will be the new numReads and numWrites because in the last message at the end it expects to read a boolean - in line 424 of filecopy.cpp there is: msg.read(ok).  The problems are
1) new dfuserver with old ftslave: the boolean "ok" field  will be mistaken for the numreads.  I believe dfuserver will crash as there would be enough data to read both numreads+numwrites+ok.   2) old dfuserver with new ftslave: ftslave deserializes numreads which is read as the "ok" field.

The current code attempts to negotiate the versions so that mismatched ftlave and dfuserver can work together.

Please let me know if there is a simpler way.

Reply:
>I can't assume that any remaining data will be the new numReads and numWrites because in the last message at the end it expects to read a boolean - in line 424 of filecopy.cpp there is: msg.read(ok).

ok, not if anything remaining, but if > than your bytesRequired... then it has to be >= version 2, and is guaranteed to be compatible. Any future version (>2) can conditionally deserialize based on version.

So I think reference should be removed, version left in for future compatibility (but not actually needed yet):

```
MemoryBuffer & OutputProgress::deserializeExtra(MemoryBuffer & in, unsigned version)
{
    if (in.remaining())
    {
        // version 1
        in.read(hasCompressed);
        if (hasCompressed)
            in.read(compressedPartSize);

        constexpr unsigned bytesRequired=sizeof(numWrites)+sizeof(numReads);
        if (in.remaining()>=bytesRequired) // must be version >=2
        {
            // NB: server has sent clients it's version and guaranteed this meta info to be compatible.
            // version 2 
            in.read(numWrites).read(numReads);
            // if version >2 can conditionally deserialize more
        }
    }
    return in;
}
```